File: scripts/cubemap2fisheye.py
#!/home/m0rtzz/Program_Files/anaconda3/envs/py38/bin/python3

import numpy as np
import math
import os
import datetime
import cv2
import argparse


def get_spherical_coordinates(FishSize):
    """
    基于输出图像的尺寸，确定输出图像上每个点的极坐标(r, phi)
    Finds spherical coordinates on the output image
    :param output_height: height of output image
    :param output_width: width of output image
    :return: two matrices that contain spherical coordinates
    for all pixels of the output image
    """
    center = int(FishSize/2)
    cc = (center, center)

    y = np.arange(0, FishSize, 1)  # 返回一个有终点和起点的固定步长的排列（等差数组）
    x = np.arange(0, FishSize, 1)

    xx, yy = np.meshgrid(y, x)  # 将y、x中每个数据排列组合生成多个点，将各个点的x坐标放入xx中，y坐标放入yy中

    bias = np.ones((FishSize, FishSize))*cc[0]  # 生成元素全为cc[0]的矩阵

    xx = np.subtract(xx, bias)  # 将横坐标范围从0~output_height变为+-output_height/2
    yy = np.subtract(yy, bias)
    xx = np.divide(xx, bias)     # 横纵坐标标准化
    xx[:, -1] = 1
    yy = np.divide(yy, -bias)
    yy[-1, :] = -1

    r = np.sqrt(xx**2 + yy**2)  # 每个点的极坐标距离，并消除距离大于1的点
    r[r > 1] = np.nan
    r[r < 0] = 0

    phi = np.zeros((FishSize, FishSize))   # 计算每个点的极坐标角度
    phi[:center, center:] = np.arcsin(np.divide(yy[:center, center:], r[:center, center:]))
    phi[:, :center] = np.pi - np.arcsin(np.divide(yy[:, :center], r[:, :center]))
    phi[center+1:, center:] = 2*np.pi + np.arcsin(np.divide(yy[center+1:, center:], r[center+1:, center:]))
    phi[cc[0], cc[1]] = 0

    return r, phi


def spherical_to_cartesian(r, phi, fov):
    """
    Transforms spherical coordinates to cartesian
    :param r: matrix with computed pixel heights
    :param phi: matrix with computed pixel angles
    :param fov: desired field of view
    :return: x,y,z cartesian coordinates
    """
    theta = r*fov/2
    x = np.sin(theta)*np.cos(phi)
    y = np.sin(theta)*np.sin(phi)
    z = np.cos(theta)

    return x, y, z


def get_face(x, y, z):
    """
    根据三维笛卡尔坐标，确定该点位于立方体哪个表面上
    Finds which face of a cube map a 3D vector with origin
    at the center of the cube points to
    :param x, y, z: cartesian coordinates
    :return: string that indicates the face
    """

    max_axis = max(abs(x), abs(y), abs(z))

    if math.isclose(max_axis, abs(x)):  # math.isclose 两个数是否绝对/相对接近
        return 'right' if x < 0 else 'left'
    elif math.isclose(max_axis, abs(y)):
        return 'bottom' if y < 0 else 'top'
    elif math.isclose(max_axis, abs(z)):
        return 'back' if z < 0 else 'front'


def raw_face_coordinates(face, x, y, z):
    """
    Finds u,v coordinates (image coordinates) for a given
    3D vector
    :param face: face where the vector points to
    :param x, y, z: vector cartesian coordinates
    :return: uv image coordinates
    """
    if face == 'left':
        u = z
        v = -y
        ma = abs(x)
    elif face == 'right':
        u = -z
        v = -y
        ma = abs(x)
    elif face == 'bottom':
        u = -x
        v = -z
        ma = abs(y)
    elif face == 'top':
        u = -x
        v = z
        ma = abs(y)
    elif face == 'back':
        u = x
        v = y
        ma = abs(z)
    elif face == 'front':
        u = -x
        v = -y
        ma = abs(z)
    else:
        raise Exception('Tile ' + face + 'does not exist')

    return (u/ma + 1)/2, (v/ma + 1)/2


def normalized_coordinates(face, x, y, PicSize):
    """
    Finds coordinates on the 2D cube map image of a 3D
    vector
    :param face: face where a 3D vector points to
    :param x, y: image coordinates
    :param n: tiles size
    :return: coordinates on the 2D cube map image
    """
    # 首先获取该表面的原点坐标
    face_origin = {
        'left': (0, PicSize),
        'front': (PicSize, PicSize),
        'right': (2*PicSize, PicSize),
        'back': (3*PicSize, PicSize),
        'top': (PicSize, 0),
        'bottom': (PicSize, 2*PicSize),
    }
    tile_origin_coords = face_origin.get(face)

    tile_x = math.floor(x*PicSize)    # math.floor 向下取整函数
    tile_y = math.floor(y*PicSize)

    if tile_x < 0:
        tile_x = 0
    elif tile_x >= PicSize:
        tile_x = PicSize-1
    if tile_y < 0:
        tile_y = 0
    elif tile_y >= PicSize:
        tile_y = PicSize-1

    x_cubemap = tile_origin_coords[0] + tile_x
    y_cubemap = tile_origin_coords[1] + tile_y

    return x_cubemap, y_cubemap


def cube2fisheye(picture_group, PicSize, FishSize, FOV):
    lPic, rPic, tPic, bPic, fPic = picture_group
    cubemap = np.zeros((3*PicSize, 3*PicSize, 3))
    cubemap[PicSize:2*PicSize, 0:PicSize] = lPic
    cubemap[PicSize:2*PicSize, PicSize:2*PicSize] = fPic
    cubemap[PicSize:2*PicSize, 2*PicSize:3*PicSize] = rPic
    cubemap[0:PicSize, PicSize:2*PicSize] = tPic
    cubemap[2*PicSize:3*PicSize, PicSize:2*PicSize] = bPic

    fisheye_picture = np.zeros((FishSize, FishSize, 3))
    fov = FOV*np.pi/180
    r, phi = get_spherical_coordinates(FishSize)
    x, y, z = spherical_to_cartesian(r, phi, fov)

    for row in range(0, FishSize):
        for column in range(0, FishSize):
            if np.isnan(r[row, column]):    # 将输出图像平面上极坐标距离超过1的点设置为黑色
                fisheye_picture[row, column, :] = 0
            # 对于极坐标距离在1以内的点
            else:
                # 首先确定该点对应的三维坐标指向哪个立方体表面
                face = get_face(x[row, column],
                                y[row, column],
                                z[row, column])
                # 然后确定该点在该立方体表面上的uv坐标
                u, v = raw_face_coordinates(face,
                                            x[row, column],
                                            y[row, column],
                                            z[row, column])
                # 最后获取标准化的uv坐标，锁定输出图像上坐标为(row,column)的点对应着立方体哪个表面上的哪个点
                x_cubemap, y_cubemap = normalized_coordinates(face,
                                                              u,
                                                              v,
                                                              PicSize)
                # 将原图像素色彩值转移到输出图像上
                fisheye_picture[row, column, :] = cubemap[y_cubemap, x_cubemap, :]

    return fisheye_picture


def main(input_images):
    PicSize = 1024  # 立方体贴图每个面的尺寸
    FishSize = 1344  # 鱼眼图像尺寸
    FOV = 196  # 鱼眼视角
    output_path = './images/my_images/fisheye_transformation/cubemap2fisheye/'
    output_file_name = datetime.datetime.now().strftime('%Y-%m-%d_%H:%M:%S') + ".png"
    # 加载输入图像
    input_images = sorted(input_images, key=lambda x: (
        'LEFT' in x, 'RIGHT' in x, 'TOP' in x, 'BOTTOM' in x, 'FRONT' in x), reverse=True)
    print(input_images)
    lPic = cv2.imread(input_images[0])
    rPic = cv2.imread(input_images[1])
    tPic = cv2.imread(input_images[2])
    bPic = cv2.imread(input_images[3])
    fPic = cv2.imread(input_images[4])
    # 生成鱼眼图像
    fisheye_picture = cube2fisheye([lPic, rPic, tPic, bPic, fPic], PicSize, FishSize, FOV)

    # 保存鱼眼图像
    cv2.imwrite(output_path + output_file_name, fisheye_picture)
    if os.path.exists(output_path + output_file_name):
        print("转换成功")
        cv2.namedWindow("Press ESC to exit", cv2.WINDOW_NORMAL | cv2.WINDOW_GUI_NORMAL)
        cv2.resizeWindow("Press ESC to exit", 800, 600)  # 设置窗口大小为 800x600
        mat = cv2.imread(output_path + output_file_name)
        cv2.imshow("Press ESC to exit", mat)
        while True:
            if cv2.waitKey(1) == 27:  # 按下esc键的ASCII码为27
                break


if __name__ == '__main__':
    parser = argparse.ArgumentParser(description='Generate fisheye image from cube map')
    parser.add_argument('input_images', type=str, nargs=5, help='input cube map images')

    args = parser.parse_args()
    main(args.input_images)

File: scripts/cubemap2fisheye.cc
/**
 * @file cubemap2fisheye.cc
 * @brief  ./cubemap2fisheye.py c++重写
 * @author M0rtzz E-mail : m0rtzz@outlook.com
 * @version 1.0
 * @date 2024-02-03
 *
 */

#include <Eigen/Dense>
#include <algorithm>
#include <cmath>
#include <filesystem>
#include <iostream>
#include <opencv2/opencv.hpp>
#include <string>
#include <tuple>
#include <vector>

using namespace std;
using namespace cv;

/**
 * @brief  将字符串转换为小写
 * @param  str
 * @return string
 */
string toLower(const string &str)
{
    string lowerStr = str;
    transform(str.begin(), str.end(), lowerStr.begin(),
              [](char c)
              { return c >= 'A' && c <= 'Z' ? c + 32 : c; });
    return lowerStr;
}

/**
 * @brief  比较函数，按照特定顺序排序
 * @param  a
 * @param  b
 * @return true
 * @return false
 */
bool compareImageNames(const string &a, const string &b)
{
    vector<string> order = {"left", "right", "top", "bottom", "front"};

    string lower_a = toLower(a);
    string lower_b = toLower(b);

    vector<int> priorityA(5, 0), priorityB(5, 0);

    for (int i = 0; i < order.size(); ++i)
    {
        if (lower_a.find(order[i]) != string::npos)
        {
            priorityA[i] = 1;
        }
        if (lower_b.find(order[i]) != string::npos)
        {
            priorityB[i] = 1;
        }
    }
    return lexicographical_compare(priorityA.begin(), priorityA.end(), priorityB.begin(), priorityB.end(), greater<int>());
}

/**
 * @brief  获取当前时间并格式化字符串
 * @return string
 */
string currentTimeFormatted()
{
    auto now = chrono::system_clock::now();
    auto in_time_t = chrono::system_clock::to_time_t(now);

    stringstream ss;
    ss << put_time(localtime(&in_time_t), "%Y-%m-%d_%H:%M:%S");
    return ss.str();
}

/**
 * @brief  生成基于图像尺寸的球面坐标
 * @param  fish_size -- 鱼眼图像的尺寸
 * @return pair<Eigen::MatrixXd, Eigen::MatrixXd> -- 包含两个矩阵的pair，分别代表r和phi坐标
 */
pair<Eigen::MatrixXd, Eigen::MatrixXd> getSphericalCoordinates(int fish_size)
{
    int center = fish_size / 2;
    Eigen::MatrixXd xx = Eigen::MatrixXd::Zero(fish_size, fish_size);
    Eigen::MatrixXd yy = Eigen::MatrixXd::Zero(fish_size, fish_size);

    for (int i = 0; i < fish_size; ++i)
    {
        for (int j = 0; j < fish_size; ++j)
        {
            xx(i, j) = j - center;
            yy(i, j) = center - i; // 注意Y轴在图像中向下是正方向
        }
    }

    xx /= center;
    yy /= center;

    Eigen::MatrixXd r = (xx.array().square() + yy.array().square()).sqrt();
    r = r.unaryExpr([](double v)
                    { return (v <= 1.0) ? v : nan(""); });

    Eigen::MatrixXd phi = Eigen::MatrixXd::Zero(fish_size, fish_size);
    for (int i = 0; i < fish_size; ++i)
    {
        for (int j = 0; j < fish_size; ++j)
        {
            if (!isnan(r(i, j)) && r(i, j) > 0)
            {
                phi(i, j) = atan2(yy(i, j), xx(i, j));
            }
        }
    }

    return {r, phi};
}

/**
 * @brief  将球面坐标转换为笛卡尔坐标
 * @param  r -- 球面坐标中的r矩阵
 * @param  phi -- 球面坐标中的phi矩阵
 * @param  FOV -- 视场(Field of View)
 * @return tuple<Eigen::MatrixXd, Eigen::MatrixXd, Eigen::MatrixXd> -- 包含x, y, z坐标矩阵的元组
 */
tuple<Eigen::MatrixXd, Eigen::MatrixXd, Eigen::MatrixXd> sphericalToCartesian(const Eigen::MatrixXd &r, const Eigen::MatrixXd &phi, double FOV)
{
    Eigen::MatrixXd theta = r * FOV / 2;
    Eigen::MatrixXd x = theta.unaryExpr([](double v)
                                        { return sin(v); })
                            .array() *
                        phi.unaryExpr([](double v)
                                      { return cos(v); })
                            .array();
    Eigen::MatrixXd y = theta.unaryExpr([](double v)
                                        { return sin(v); })
                            .array() *
                        phi.unaryExpr([](double v)
                                      { return sin(v); })
                            .array();
    Eigen::MatrixXd z = theta.unaryExpr([](double v)
                                        { return cos(v); });

    return {x, y, z};
}

/**
 * @brief  根据笛卡尔坐标确定点所在的立方体表面
 * @param  x
 * @param  y
 * @param  z
 * @return string -- 点所在的立方体表面名称
 */
string getFace(double x, double y, double z)
{
    double max_axis = max({abs(x), abs(y), abs(z)});
    if (abs(x) == max_axis)
    {
        return x < 0 ? "right" : "left";
    }
    else if (abs(y) == max_axis)
    {
        return y < 0 ? "bottom" : "top";
    }
    else
    {
        return z < 0 ? "back" : "front";
    }
}

/**
 * @brief  计算立方体表面上的u, v坐标
 * @param  face
 * @param  x
 * @param  y
 * @param  z
 * @return pair<double, double> -- 立方体表面上的u, v坐标
 */
pair<double, double> rawFaceCoordinates(const string &face, double x, double y, double z)
{
    double u, v, ma;
    if (face == "left" || face == "right")
    {
        u = (face == "left") ? z : -z;
        v = -y;
        ma = abs(x);
    }
    else if (face == "bottom" || face == "top")
    {
        u = -x;
        v = (face == "bottom") ? -z : z;
        ma = abs(y);
    }
    else
    { // "back" or "front"
        u = (face == "front") ? -x : x;
        v = -y;
        ma = abs(z);
    }
    return {(u / ma + 1) / 2, (v / ma + 1) / 2};
}

/**
 * @brief  将立方体表面上的坐标转换为2D图像上的坐标
 * @param  face -- 点所在的立方体表面名称
 * @param  u -- 立方体表面上的u坐标
 * @param  v -- 立方体表面上的v坐标
 * @param  pic_size -- 图像尺寸
 * @return pair<int, int> -- 2D图像上的坐标
 */
pair<int, int> normalizedCoordinates(const string &face, double u, double v, int pic_size)
{
    map<string, pair<int, int>> face_origin = {
        {"left", {0, pic_size}},
        {"front", {pic_size, pic_size}},
        {"right", {2 * pic_size, pic_size}},
        {"back", {3 * pic_size, pic_size}},
        {"top", {pic_size, 0}},
        {"bottom", {pic_size, 2 * pic_size}}};

    auto [face_x, face_y] = face_origin[face];
    int x = floor(u * pic_size);
    int y = floor(v * pic_size);

    x = clamp(x, 0, pic_size - 1);
    y = clamp(y, 0, pic_size - 1);

    return {face_x + x, face_y + y};
}

/**
 * @brief  将立方体映射图像转换为鱼眼图像
 * @param  picture_group -- 包含立方体六个面图像的向量
 * @param  pic_size -- 立方体贴图每个面的尺寸
 * @param  fish_size -- 鱼眼图像的尺寸
 * @param  FOV -- 视场(Field of View)
 * @return Mat -- 鱼眼图像
 */
Mat cube2Fisheye(const vector<Mat> &picture_group, int pic_size, int fish_size, double FOV)
{
    Mat cubemap(3 * pic_size, 3 * pic_size, CV_8UC3, Scalar(0, 0, 0));
    // 将输入的图像复制到立方体贴图的相应位置
    picture_group[0].copyTo(cubemap(Rect(0, pic_size, pic_size, pic_size)));            // left
    picture_group[1].copyTo(cubemap(Rect(2 * pic_size, pic_size, pic_size, pic_size))); // right
    picture_group[2].copyTo(cubemap(Rect(pic_size, 0, pic_size, pic_size)));            // top
    picture_group[3].copyTo(cubemap(Rect(pic_size, 2 * pic_size, pic_size, pic_size))); // bottom
    picture_group[4].copyTo(cubemap(Rect(pic_size, pic_size, pic_size, pic_size)));     // front

    auto [r, phi] = getSphericalCoordinates(fish_size);
    auto [x, y, z] = sphericalToCartesian(r, phi, FOV * M_PI / 180);

    Mat fisheye_picture = Mat::zeros(fish_size, fish_size, CV_8UC3);
    for (int i = 0; i < fish_size; ++i)
    {
        for (int j = 0; j < fish_size; ++j)
        {
            if (isnan(r(i, j)))
            {
                continue;
            }

            string face = getFace(x(i, j), y(i, j), z(i, j));
            auto [u, v] = rawFaceCoordinates(face, x(i, j), y(i, j), z(i, j));
            auto [x_cubemap, y_cubemap] = normalizedCoordinates(face, u, v, pic_size);

            fisheye_picture.at<Vec3b>(i, j) = cubemap.at<Vec3b>(y_cubemap, x_cubemap);
        }
    }

    return fisheye_picture;
}

int main(int argc, char **argv)
{
    if (argc != 6)
    {
        cerr << "Usage: cube2Fisheye <image> <image> <image> <image> <image>" << endl;
        return EXIT_FAILURE;
    }

    vector<string> image_paths;
    for (int i = 1; i < argc; ++i)
    {
        image_paths.push_back(argv[i]);
    }

    sort(image_paths.begin(), image_paths.end(), compareImageNames);

    vector<Mat> input_images;
    for (const auto &imagePath : image_paths)
    {
        input_images.push_back(imread(imagePath));
    }

    int pic_size = 1024;  // 立方体贴图每个面的尺寸
    int fish_size = 1344; // 鱼眼图像尺寸
    double FOV = 196;     // 鱼眼视角

    Mat fisheye_picture = cube2Fisheye(input_images, pic_size, fish_size, FOV);

    string output_file_name = "./images/my_images/fisheye_transformation/cubemap2fisheye/" + currentTimeFormatted() + ".png";
    imwrite(output_file_name, fisheye_picture);

    // cout << "\n转换成功，文件保存为: " << output_file_name << endl;

    return EXIT_SUCCESS;
}
